File: cpp/HybridPdfPage.cpp
#include "HybridPdfPage.hpp"
#include "HybridPdfAnnotation.hpp"
#include "HybridPdfPainter.hpp"

namespace margelo::nitro::pdfeditor {

using namespace PoDoFo;

double HybridPdfPage::getWidth() {
  std::lock_guard<std::mutex> lock(*_mutex);
  return _page->GetRect().Width;
}

double HybridPdfPage::getHeight() {
  std::lock_guard<std::mutex> lock(*_mutex);
  return _page->GetRect().Height;
}

double HybridPdfPage::getIndex() {
  std::lock_guard<std::mutex> lock(*_mutex);
  // GetIndex() is already the 0-based index (confirmed against the podofo
  // repo's own working JNI wrapper, src/wrapper/podofo_jni.cpp, which uses
  // this same call) — no off-by-one adjustment needed.
  return static_cast<double>(_page->GetIndex());
}

std::shared_ptr<HybridPdfPainterSpec> HybridPdfPage::createPainter() {
  std::lock_guard<std::mutex> lock(*_mutex);
  return std::make_shared<HybridPdfPainter>(_doc, _mutex, _page);
}

double HybridPdfPage::getAnnotationCount() {
  std::lock_guard<std::mutex> lock(*_mutex);
  return static_cast<double>(_page->GetAnnotations().GetCount());
}

std::shared_ptr<HybridPdfAnnotationSpec> HybridPdfPage::getAnnotationAt(
    double index) {
  std::lock_guard<std::mutex> lock(*_mutex);
  auto& annotation =
      _page->GetAnnotations().GetAnnotAt(static_cast<unsigned>(index));
  return std::make_shared<HybridPdfAnnotation>(_doc, _mutex, &annotation);
}

std::shared_ptr<HybridPdfAnnotationSpec> HybridPdfPage::createAnnotation(
    PdfAnnotationType annotationType, double x, double y, double width,
    double height) {
  std::lock_guard<std::mutex> lock(*_mutex);
  auto& annotation = _page->GetAnnotations().CreateAnnot(
      toPodofoAnnotationType(annotationType), Rect(x, y, width, height));
  return std::make_shared<HybridPdfAnnotation>(_doc, _mutex, &annotation);
}

std::vector<PdfTextEntry> HybridPdfPage::extractText(
    const std::optional<std::string>& pattern) {
  std::lock_guard<std::mutex> lock(*_mutex);
  std::vector<PoDoFo::PdfTextEntry> podofoEntries;
  _page->ExtractTextTo(podofoEntries, pattern.value_or(std::string()));

  std::vector<PdfTextEntry> entries;
  entries.reserve(podofoEntries.size());
  for (const auto& entry : podofoEntries) {
    entries.emplace_back(entry.Text, entry.X, entry.Y, entry.Length);
  }
  return entries;
}

}  // namespace margelo::nitro::pdfeditor

File: android/src/main/java/com/margelo/nitro/pdfeditor/HybridPdfPage.kt
package com.margelo.nitro.pdfeditor

import com.facebook.proguard.annotations.DoNotStrip
import com.podofo.android.PdfPage as PodofoPage

/** [lock] is the same one shared by the owning HybridPdfDocument — see its class doc for why. */
@DoNotStrip
class HybridPdfPage(internal val native: PodofoPage, private val lock: Any) : HybridPdfPageSpec() {
  override val width: Double
    get() = synchronized(lock) { native.width }

  override val height: Double
    get() = synchronized(lock) { native.height }

  override val index: Double
    get() = synchronized(lock) { native.index.toDouble() }

  override fun getRotation(): Double {
    return synchronized(lock) { native.rotation.toDouble() }
  }

  override fun setRotation(rotation: Double) {
    synchronized(lock) { native.rotation = rotation.toInt() }
  }

  override fun getMediaBox(): PdfRect {
    return synchronized(lock) {
      val box = native.mediaBox
      PdfRect(box[0], box[1], box[2], box[3])
    }
  }

  override fun setMediaBox(x: Double, y: Double, width: Double, height: Double) {
    synchronized(lock) { native.setMediaBox(x, y, width, height) }
  }

  override fun getCropBox(): PdfRect {
    return synchronized(lock) {
      val box = native.cropBox
      PdfRect(box[0], box[1], box[2], box[3])
    }
  }

  override fun setCropBox(x: Double, y: Double, width: Double, height: Double) {
    synchronized(lock) { native.setCropBox(x, y, width, height) }
  }

  override fun moveTo(newIndex: Double): Boolean {
    return synchronized(lock) { native.moveTo(newIndex.toInt()) }
  }

  override fun createPainter(): HybridPdfPainterSpec {
    return synchronized(lock) { HybridPdfPainter(native, lock) }
  }

  override fun getAnnotationCount(): Double {
    return synchronized(lock) { native.annotationCount.toDouble() }
  }

  override fun getAnnotationAt(index: Double): HybridPdfAnnotationSpec {
    return synchronized(lock) { HybridPdfAnnotation(native.getAnnotationAt(index.toInt()), lock) }
  }

  override fun createAnnotation(
    annotationType: PdfAnnotationType,
    x: Double,
    y: Double,
    width: Double,
    height: Double,
  ): HybridPdfAnnotationSpec {
    return synchronized(lock) {
      HybridPdfAnnotation(
        native.createAnnotation(annotationType.toPodofoName(), x, y, width, height),
        lock,
      )
    }
  }

  override fun extractText(pattern: String?): Array<PdfTextEntry> {
    return synchronized(lock) {
      val entries = if (pattern != null) native.extractText(pattern) else native.extractText()
      entries.map { PdfTextEntry(it.text, it.x, it.y, it.length) }.toTypedArray()
    }
  }
}
